src/controllers/midi/midisourceclock.cpp

```diff
@@ -0,0 +1,131 @@
+#include "controllers/midi/midimessage.h"
+#include "controllers/midi/midisourceclock.h"
+#include "util/math.h"
+
+bool MidiSourceClock::handleMessage(unsigned char status) {
+    // TODO(owen): We need to support MIDI_CONTINUE.
+    switch (status) {
+    case MIDI_START:
+        start();
+        return true;
+    case MIDI_STOP:
+        stop();
+        return true;
+    case MIDI_TIMING_CLK:
+        pulse();
+        return true;
+    default:
+        return false;
+    }
+}
+
+void MidiSourceClock::start() {
+    // Treating MIDI_START as the first downbeat is standard practice:
+    // http://www.blitter.com/%7Erusstopia/MIDI/%7Ejglatt/tech/midispec/seq.htm
+    m_bRunning = true;
+    m_iFilled = 0;
+    m_iRingBufferPos = 0;
+}
+
+void MidiSourceClock::stop() {
+    m_bRunning = false;
+}
+
+void MidiSourceClock::pulse() {
+    // Update the ring buffer and calculate new bpm.  Update the last beat time
+    // if we are on a beat.
+
+    if (!m_bRunning) {
+        qDebug()
+                << "MidiSourceClock: Got clock pulse but not started, starting now.";
+        start();
+    }
+
+    // Ringbuffer filling.
+    // TODO(owen): We should have a ringbuffer convenience class.
+    const mixxx::Duration lastPulseTime = m_pClock->now();
+    m_pulseRingBuffer[m_iRingBufferPos] = lastPulseTime;
+    m_iRingBufferPos = (m_iRingBufferPos + 1) % kRingBufferSize;
+    if (m_iFilled < kRingBufferSize) {
+        ++m_iFilled;
+    }
+
+    // If this pulse is a beat mark, record it, even if we have very few samples.
+    if (m_iRingBufferPos % kPulsesPerQuarter == 0) {
+        m_lastBeatTime = lastPulseTime;
```

Comment:
It is better to take the previous m_lastBeatTime into account and add the calculated m_dBpm from below.